### DISCUSSION TITLE
Null Operator - DCSP for variables

### DISCUSSION THREAD

**Discussion by gadavidd** [chocoteam/choco-solver#1031] [General]
I am reading about DCSP (Dynamic Constraints Satisfaction), in particular focusing on Mittal's 1990 work (initial reference), and thought about the topic with Choco-solver. Taking into consideration the DCSP implementations and @cprudhom and @jgFages discussions in 2014-15 (see #232), I realized that the focus of the implementation was on enabling and disabling Constraints in Choco. However, for changes with a focus on variables following the DCSP aspects in Choco, it would not be possible. Is this true? I think about this because of an implementation I recently discussed with @TramontinLSS. 

Consider including a null (`na` - not applicable) operator. This operator would have the function of allowing a variable to take no value from the domain. In this way, the variable will be disabled in the model by this operator and will be irrelevant to the solution. Have you seen any implementation of this type? It is very useful in cases of Table Constraints and Hybrid Tables.

For example, in the case where we have variables `A=[1,2,3]; B=[4,5,6]; C=[7,8,9]` and `D=[10,11,12]`

Consider the following Table Constraints:

```java       
		tuples.add(eq(1),eq(4),eq(9));
		tuples.add(eq(3),eq(5),eq(8));
```
The solutions will be in:

           `A=[1,3]`, `B=[4,5]`, `C=[8,9]`, `D=[10,11,12]`

But if by chance it is convenient that `C` is definitely not part of the solution in a certain scenario, for example, when `D=10`:
```java 
		tuples.add(eq(2),eq(5),na(),eq(10))
```
C would be inactive if the domain `D` equals 10, so `C` would not be part of the final solution:

		        A=[2], B=[5], D=[10]

This model seems of little relevance in a model as small as the one represented above. Even in the bibliography, the null value could be included in the domain of variables, or else the smallest value of the domain could be considered for `C`. However, we have cases of extremely giant models to which the addition of this operator would be of extreme importance. Have you considered something like this? I am trying to work on this implementation in Choco.

In summary, the null operator would signal which variable does not need to receive a value to solve the problem (it would be the inverse of `any()`), becoming an irrelevant variable that indicates that effort should not be spent during the solution.

**Comment by cprudhom** [comment]
Hi @gadavidd just to say that I won't be able to discuss your point before a few days/weeks. But, it's in my stack

**Comment by cprudhom** [comment]
You draw a parallel with the constraints that can be removed from a model.
First, you should keep in mind that choco was designed in the first place to solve one (immutable) problem at a time.

## Adding/removing a constraint
There are two ways to remove a constraint `c` from a Choco's Model:
On the one hand, with a call to `model.postTemp(c);` a constraint is added temporarily  during the solving phare (at a node `n`) and automatically removed upon backtrack.
On the other hand, one can still call `model.unpost(c)` during the solving phase, but it should be followed by a call to `solver.restart()` to restore all values filtered by `c` in the current branch.

And a constraint is a function that removes values from variables' domain. Unposting a constraint _relax_ the model.

## Relaxing a variable

If I'm correct:
1. you  want to allow a variable to take no value in its domain and
2. you focus on hybrid table constraint,

right?

To me, that's why `any()` exists, to mention that a variable can take any value in its domain.
If you don't care about the value of `C` (in your example) why not defining a search strategy that manage _active_ variables dynamically, the remaining ones could be left unassigned?
